### ISSUE TITLE
Clarification on when is it safe to call vkDestroySwapchainKHR

### ISSUE THREAD

**Issue by plasmacel** [KhronosGroup/Vulkan-Docs#1678]
[VK_KHR_swapchain] @cubanismo @ianelliottus

Even after the latest changes in the wording, it is unclear when is it safe to call `vkDestroySwapchainKHR`, if ever.

> `vkQueueWaitIdle` is equivalent to having submitted a valid fence to every previously executed **queue submission command that accepts a fence**, then waiting for all of those fences to signal using `vkWaitForFences` with an infinite `timeout` and `waitAll` set to `VK_TRUE`.

and

> `vkDeviceWaitIdle` is equivalent to calling `vkQueueWaitIdle` for all queues owned by device.

make clear that `vkQueueWaitIdle` and `vkDeviceWaitIdle` don't cover pending present operations, since `vkQueuePresentKHR` doesn't accept a fence. This means that alone on their own, in the absence of any other guarantee they are not sufficient to make `vkDestroySwapchainKHR` safe.

The spec says about `vkDestroySwapchainKHR`, that

> The application must not destroy a swapchain until after **completion of all outstanding operations** on images that were acquired from the swapchain. `swapchain` and all associated `VkImage` handles are destroyed, and must not be acquired or used any more by the application. The memory of each `VkImage` will only be freed after that image is no longer **used by the presentation engine**. For example, if one image of the swapchain is being displayed in a window, the memory for that image may not be freed until the window is destroyed, or another swapchain is created for the window.

which still doesn't clarify what **"completion of all outstanding operations"** is and whether it covers pending present operations or not. The same applies to **"used by the presentation engine"** - does it include all previously submitted present operations?

Unfortunately, the `VK_KHR_present_wait` extension also doesn't solve the issue, since it doesn't work on expired swapchains, it cannot be used for swapchain resizing.

What does guarantee that `vkDestroySwapchainKHR` is safe to call?

**Comment by stonesthrow** [comment]
ACK. We will discuss a follow up action this week.

**Comment by r-potter** [comment]
This is a spec bug that requires an extension to properly fix. We have such an upcoming extension planned that should resolve it. 

Until that solution becomes available, we recommend continuing to use `vkQueueWaitIdle`.

**Comment by krOoze** [comment]
How does a new extension fix a problem for apps that do not use said extension?

**Comment by plasmacel** [comment]
It would more than great if the safety of `vkDestroySwapchainKHR` could be guaranteed by waiting a fence, without the need of  `vkQueueWaitIdle`, which would make possible to simply put the swapchain to a deferred destroy queue, just like other resources.

**Comment by plasmacel** [comment]
Regarding the extension, I agree with @krOoze. Destroying a swapchain is one of the most fundamental operations, used by the vast majority of applications. As it is, currently these applications are exposed to unspecified behavior. This raises serious concerns and would also mean that there is no standardized safe way to handle swapchains in core Vulkan 1.2 or below.

**Comment by krOoze** [comment]
It's the oldest spec defect Issue (https://github.com/KhronosGroup/Vulkan-Docs/issues/152) here, so I think it should also get some priority before it is another five years.

@plasmacel BTW which "latest changes" you refer to in the OP? I don't think there was anything significant that changed the original problem.

Anyway AIS in the other thread `vkQueueWaitIdle` does not work or would be at best very ambiguous. If that is supposed to be the way, the spec should clarify. Also should perhaps spell out in the spec the semaphore state inferrence through the swapchain image indexes.

**Comment by cubanismo** [comment]
It's not clarified to work in the spec because it isn't guaranteed to work. We're saying that if it happens to work for use cases you're aware of, then feel free to keep using it.. The spec is broken. The drivers based on that spec are, as a result, broken as well. Nothing we clarify or change in the spec is going to fix those existing drivers, so we're writing an extension that fixes the spec. I'm not clear how this could be fixed any other way.

**Comment by krOoze** [comment]
@cubanismo Ok. What do the "broken" drivers actually do then? Why not inquire them, and retrospectively specify some highest common multiple of that behavior, in order to at least sanctify the current status quo and make the "broken" but working apps that run on "broken" but working drivers valid.

**Comment by Mrkol** [comment]
I know nothing of how present engines work, but a side-effect of this issue came up when I tried to write a generalized wrapper for asynchronous vulkan operations. As far as I can tell, currently, the only async call in vulkan that provides no way to wait for completion is `vkQueuePresentKhr`, which in turns makes it impossible to safely manage swapchains: we can never be sure whether it is in use or not.
Is there some specific presentation engine implementation detail that prevents the spec from requiring the driver to signal a fence when the presentation completes? As far as I can tell, this would allow for the issue to be fixed by then specifying a swapchain to be safe to delete when all acquired images have been presented to completion. Although preferably there should be a way to cancel an acquired image without presentation so that window resizes can happen without rending a frame at incorrect resolution when an acquisition completes with a `suboptimal` return code.

**Comment by krOoze** [comment]
@Rua If you mean `VK_KHR_present_wait`, then yes.

**Comment by plasmacel** [comment]
@krOoze Unfortunately, the VK_KHR_present_wait extension doesn't really solve the issue, since it doesn't work on expired swapchains, it cannot be used for swapchain resizing.

**Comment by krOoze** [comment]
@plasmacel  I don't immediately see any such limitation. I only see it is [forbidden for retired swapchains](https://github.com/KhronosGroup/Vulkan-Docs/issues/1967), which feels like an omission though...

**Comment by ShabbyX** [comment]
The new [VK_EXT_surface_maintenance1][surface] and [VK_EXT_swapchain_maintenance1][swapchain] extensions address this issue. Please see this [blog post][blog] for a summary.

[surface]: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_surface_maintenance1.html
[swapchain]: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_swapchain_maintenance1.html
[blog]: https://www.khronos.org/blog/resolving-longstanding-issues-with-wsi

TL;DR: There is now feedback from `vkQueuePresentKHR` about when resources associated with the present operation (including the swapchain) can be destroyed.

**Comment by zakarumych** [comment]
> TL;DR: There is now feedback from vkQueuePresentKHR about when resources associated with the present operation (including the swapchain) can be destroyed.

I've implemented this to get rid of `vkDeviceWaitIdle` before destroying swapchain and semaphores used for acquiring and presentation.
However I had to include a fence in every `vkQueuePresentKHR` call because I can never know which might be the last.
Could this create any noticeable overhead?

**Comment by ShabbyX** [comment]
> > TL;DR: There is now feedback from vkQueuePresentKHR about when resources associated with the present operation (including the swapchain) can be destroyed.
> 
> I use this to get rid of `vkDeviceWaitIdle` before destroying swapchain and semaphores used for acquiring and presentation. However I had to include a fence in every `vkQueuePresentKHR` call because I can never know which might be the last. Could this create any noticeable overhead?

It's a once-per-frame operation, so that would be highly unlikely; you can always measure!

The way I use it is to recycle the fences. With appropriate CPU throttling in place (so your submissions don't get too far ahead of GPU/presentation), you'll really only need like 3 fences to cycle through.

**Comment by ShabbyX** [comment]
> > > > TL;DR: There is now feedback from vkQueuePresentKHR about when resources associated with the present operation (including the swapchain) can be destroyed.
> > > 
> > > 
> > > I use this to get rid of `vkDeviceWaitIdle` before destroying swapchain and semaphores used for acquiring and presentation. However I had to include a fence in every `vkQueuePresentKHR` call because I can never know which might be the last. Could this create any noticeable overhead?
> > 
> > 
> > It's a once-per-frame operation, so that would be highly unlikely; you can always measure!
> > The way I use it is to recycle the fences. With appropriate CPU throttling in place (so your submissions don't get too far ahead of GPU/presentation), you'll really only need like 3 fences to cycle through.
> 
> Yeah, I do the throttling and recycle fences used for presentation when the oldest one becomes signaled, so I only get at most one more than limit of concurrent fenced submissions.

:+1:  Then I wouldn't really worry about a performance impact here.
